### PR TITLE
Main refactoring

### PR DESCRIPTION
**Pull request by Avielyo10**
# Description

* Use some functions for maintainability
* Remove unneeded channels & goroutines
* Register consumer closing in a graceful manner.
* Anonymouse consumers.
* Handle nil kafka config in the the service creation.
* Close() is a part of the ConsumerService

## Type of change

What is it?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [ ] Tests update
- [x] Refactor

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I run `go fmt ./...` to check that my code is properly formatted
- [x] I run `go vet ./...` to check that my code is free of common Go style mistakes

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -40,33 +41,28 @@ func initDependencies() {
...
-		"DefaultOSTreeRef":         cfg.DefaultOSTreeRef,
-		"InventoryURL":             cfg.InventoryConfig.URL,
-		"PlaybookDispatcherConfig": cfg.PlaybookDispatcherConfig.URL,
-		"TemplatesPath":            cfg.TemplatesPath,
-		"DatabaseType":             cfg.Database.Type,
-		"DatabaseName":             cfg.Database.Name,
-		"IsKafkaEnabled":           cfg.KafkaConfig != nil,
-		"FDOHostURL":               cfg.FDO.URL,
-	}).Info("Configuration Values")
+func serveMetrics(port int) *http.Server {
```
Comment (AaronH88) [inline main.go:44]: Moving this to its own file might improve readability. Something like a `metrics.go`

Comment (AaronH88) [reply main.go:44]: I really like this idea, good job!

File: main.go [hunk 2]
```diff
@@ -76,21 +72,21 @@ func main() {
 	)
 
 	// Unauthenticated routes
-	r.Get("/", routes.StatusOK)
-	r.Get("/api/edge/v1/openapi.json", func(w http.ResponseWriter, r *http.Request) {
+	route.Get("/", routes.StatusOK)
```
Comment (AaronH88) [inline main.go:75]: Love these more explicit var names

File: main.go [hunk 3]
```diff
@@ -99,57 +95,88 @@ func main() {
...
-		sigint := make(chan os.Signal, 1)
-		signal.Notify(sigint, os.Interrupt, syscall.SIGTERM)
-		<-sigint
-		log.Info("Shutting down gracefully...")
-		if err := srv.Shutdown(context.Background()); err != nil {
-			l.LogErrorAndPanic("HTTP server (web port) shutdown failed", err)
+	server.RegisterOnShutdown(func() {
+		for _, consumer := range consumers {
+			if consumer != nil {
+				consumer.Close()
```
Comment (AaronH88) [inline main.go:111]: Very nice

File: main.go [hunk 4]
```diff
@@ -99,57 +95,88 @@ func main() {
...
 	}
+	webServer := serveWeb(cfg, consumers)
+	metricsServer := serveMetrics(cfg.MetricsPort)
 
-	<-gracefulStop
+	if cfg.KafkaConfig != nil {
+		log.Info("Starting Kafka Consumers")
+		for _, consumer := range consumers {
+			if consumer != nil {
+				go consumer.Start()
```
Comment (AaronH88) [inline main.go:173]: Very very nice!

File: main.go [hunk 5]
```diff
@@ -99,57 +95,88 @@ func main() {
...
+			}
+		}
+	}
+	<-sigint
```
Comment (AaronH88) [inline main.go:177]: Humm, can we think of a more verbose name here?

File: pkg/services/consumers.go [hunk 6]
```diff
@@ -64,25 +62,28 @@ type PlatformInsightsCreateEventPayload struct {
...
-		topic:         topic,
-	}
-	if topic == "platform.playbook-dispatcher.runs" {
-		s.consumer = s.ConsumePlaybookDispatcherRuns
-	} else if s.topic == "platform.inventory.events" {
-		s.consumer = s.ConsumeInventoryCreateEvents
-	} else {
-		log.Errorf("No consumer for topic: %s", topic)
-		return nil
+	if config != nil {
```
Comment (AaronH88) [inline pkg/services/consumers.go:65]: If we change the condition here we can improve readability a bit, and also remove some nested `if`'s

e.g. `if config == nil { return nil }`

File: main.go (hunk 5 above)
Comment (AaronH88) [reply main.go:177]: Also, where was this `sigint` channel created?

File: main.go (hunk 1 above)
Comment (Avielyo10) [reply main.go:44]: You mean separate the metrics & the web servers?

File: main.go (hunk 5 above)
Comment (Avielyo10) [reply main.go:177]: like `interruptSignal`?
Created at the beginning of `main()`

File: pkg/services/consumers.go (hunk 6 above)
Comment (Avielyo10) [reply pkg/services/consumers.go:65]: Like that!

File: main.go (hunk 1 above)
Comment (AaronH88) [reply main.go:45]: Exactly, I personally would expect to find this either under `pkg/routes/` or `pkg/services/` 
Just a thought :smile:

File: main.go (hunk 5 above)
Comment (AaronH88) [reply main.go]: Great choice

File: main.go (hunk 1 above)
Comment (Avielyo10) [reply main.go:45]: Can this take place in a new PR?

**Review by ldjebran — COMMENTED** [review]
It looks good, waiting @AaronH88 response

**Review by loadtheaccumulator — CHANGES_REQUESTED** [review]
Do all of these commits have to go in at the same time, or can they be broken into separate PRs. We've been running into breakage issues with large refactors.

I think we might need to sync more on refactor plans versus new feature now that we're in squads and might be touching the same code.

**Comment by Avielyo10** [comment]
@loadtheaccumulator, actually they are all connected in some way.

This is not a feature, only improve readability & avoid goroutines leak on main (also implement a real graceful shutdown)

**Comment by Avielyo10** [comment]
@mergify rebase

File: main.go (hunk 1 above)
Comment (AaronH88) [reply main.go:45]: Yes it can, Ok move this in a new PR.

**Comment by AaronH88** [comment]
Looks good to me, ok to merge as well as have that chat Jonathan is pointing out

**Comment by Avielyo10** [comment]
@loadtheaccumulator ?

**Pull request merged**